relentless/core/variable.py

```diff
@@ -0,0 +1,395 @@
+        for a in g:
+            if isinstance(a,DependentVariable):
+                for p,b in a.depends:
+                    # add new edge for p if not already present, otherwise append to params list
+                    if (a,b) not in g.edges:
+                        g.add_edge(a,b,params=[p])
+                    else:
+                        g.edges[a,b]['params'].append(p)
+
+        return g
+
+    def derivative(self, var):
+        """Calculates the derivative of a DependentVariable object
+        with respect to another Variable object.
+
+        Parameters
+        ----------
+        var : :py:class:`Variable`
+            The variable with respect to which to take the derivative.
+
+        Returns
+        -------
+        float
+            The calculated derivative value.
+
+        Raises
+        ------
+        RuntimeError
+            If the specified DependentVariable object has any circular dependencies.
+
+        """
+        # if var is this variable, the derivative is trivially 1.0
+        if var is self:
+            return 1.0
+
+        # get dependency graph
+        g = self._assert_acyclic(self.dependency_graph())
+
+        # if var is not in the graph, then its derivative is trivially 0.0
+        if var not in g:
+            return 0.0
+
+        # add sum of parameter derivatives to edges between objects
+        for a,b,params in g.edges.data('params'):
+            g.edges[a,b]['deriv'] = np.sum([a._derivative(p) for p in params])
+
+        # compute chain rule along all paths to the variable
+        deriv = 0.
+        paths = nx.all_simple_paths(g, source=self, target=var)
+        for path in map(nx.utils.pairwise, paths):
+            deriv += np.prod([g.edges[edge]['deriv'] for edge in path])
+        return deriv
+
+    @abc.abstractmethod
+    def _derivative(self, param):
+        pass
+
+    def _assert_variable(self, v):
+        #Checks if the dependent variable depends on another variable.
+        if not isinstance(v, Variable):
+            raise TypeError('Dependent variables can only depend on other variables.')
+        return v
+
+    def _assert_acyclic(self, g):
+        # confirm dependency graph is free of cycles
+        if not nx.is_directed_acyclic_graph(g):
+            raise RuntimeError('DependentVariable has circular dependencies.')
+        return g
+
+class UnaryOperator(DependentVariable):
+    """Abstract base class for dependent variable based on one parameter value."""
+    def __init__(self, a):
+        super().__init__(a=a)
+
+class SameAs(UnaryOperator):
+    """Unary operator for copying a variable."""
+    @property
+    def value(self):
+        return self.a.value
+
+    def _derivative(self, param):
+        if param == 'a':
+            return 1.0
+        else:
+            raise ValueError('Unknown parameter')
+
+class BinaryOperator(DependentVariable):
+    """Abstract base class for dependent variable based on two parameter values."""
```

Comment:
```suggestion
    """Abstract base class for a value that depends on two :py:class:`Variable`s."""
```

Please also document the two arguments `a` and `b`.